### PR TITLE
Fix the applyMatrix cudaFree segfault on LK+CUDA

### PR DESCRIPTION
**Pull request by maliasadi**
### Before submitting

Please complete the following checklist when submitting a PR:

- [ ] All new features must include a unit test.
      If you've fixed a bug or added code that should be tested, add a test to the
      [`tests`](../tests) directory!

- [x] All new functions and code must be clearly commented and documented.
      If you do make documentation changes, make sure that the docs build and
      render correctly by running `make docs`.

- [x] Ensure that the test suite passes, by running `make test`.

- [x] Add a new entry to the `.github/CHANGELOG.md` file, summarizing the
      change, and including a link back to the PR.

- [x] Ensure that code is properly formatted by running `make format`. 

When all the above are checked, delete everything above the dashed
line and fill in the pull request template.

------------------------------------------------------------------------------------------------------------

**Context:**
- Update Kokkos version support to 4.4.1 and enable LK+CUDA C++ tests on CI.
- Fix the Lightning-Kokkos[CUDA] C++ API `cudaFree` segfaults in `applyMatrix`.
- Update tests to check `applyMatrix` with both const and non-const matrix data
- Simplify LK measurement tests to speed-up the testing on CI by up to 5X

[sc-64601]
[sc-78848]

### CODE REVIEW

File: pennylane_lightning/core/src/simulators/lightning_kokkos/StateVectorKokkos.hpp [hunk 1]
```diff
@@ -486,8 +486,17 @@ class StateVectorKokkos final
...
+        const std::size_t n = static_cast<std::size_t>(1U) << wires.size();
+        const std::size_t n2 = n * n;
+        KokkosVector matrix_("matrix_", n2);
+
+        // Note that when copying data between different memory spaces (host !=
+        // device), Kokkos::View<Kokkos::complex*> cannot perform a deep copy of
+        // unmanaged complex numbers during initialization via its constructor.
+        // Thus, we need to explicitly deep-copy the matrix data using
+        // Kokkos::deep_copy().
+        Kokkos::deep_copy(matrix_, UnmanagedComplexHostView(matrix, n2));
```
Comment (multiphaseCFD) [inline pennylane_lightning/core/src/simulators/lightning_kokkos/StateVectorKokkos.hpp:498]: Nice catch! Thanks @maliasadi !

**Review by mlxd — COMMENTED** [review]
Thanks @maliasadi and good catch

Just 1 question

File: pennylane_lightning/core/src/simulators/lightning_kokkos/measurements/tests/Test_StateVectorKokkos_Measure.cpp [hunk 2]
```diff
@@ -251,8 +251,11 @@ TEMPLATE_TEST_CASE("Probabilities", "[Measures]", float, double) {
                          Catch::Approx(probabilities).margin(1e-6));
         }
     }
-    SECTION("21 qubits") {
-        constexpr std::size_t num_qubits = 21;
+    SECTION("14 qubits") {
+        // Reduce the number of qubits to 14 (from 21) will enable us to
+        // test the device on GPU boxes with a relatively small L0 scratch
+        // memory.
```
[the remark this answers is not in the document]
Comment (maliasadi) [reply pennylane_lightning/core/src/simulators/lightning_kokkos/measurements/tests/Test_StateVectorKokkos_Measure.cpp]: Good point! The main reason for reducing the number of qubits from 21 to 14 was to enable LK+CUDA to be tested on CIs ([this line](https://github.com/PennyLaneAI/pennylane-lightning/pull/1000/files#diff-de4a55e0a0119574998aba883c8921afd9488748013312be42c227b65fad04e2L109)). I just reverted it to double-check the L0 memory issue after [these updates](https://github.com/PennyLaneAI/pennylane-lightning/pull/1000/files#diff-27ebb5e9e920afe73766d5ea1b7cdda72729924a2bcda100b7e09b9491d32930R337) :crossed_fingers:

**Review by multiphaseCFD — APPROVED** [review]
LGTM! Nice work @maliasadi !

File: pennylane_lightning/core/src/utils/TestHelpers.hpp [hunk 3]
```diff
@@ -231,6 +235,37 @@ isApproxEqual(const Data_t *data1, const std::size_t length1,
...
+ * @tparam Data_t Floating point data-type.
+ * @param data1 StateVector data array pointer 1.
+ * @param data2 StateVector data array pointer 2.
+ * @param eps The absolute tolerance parameter.
+ * @return true Data are approximately equal.
+ * @return false Data are not approximately equal.
+ */
+template <class Data_t, class Alloc>
+inline bool
+isApproxEqual(const std::vector<Data_t, Alloc> &data1,
```
Comment (josephleekl) [inline pennylane_lightning/core/src/utils/TestHelpers.hpp:254]: Can we add a comment why this is necessary for Kokkos vectors?

**Review by josephleekl — APPROVED** [review]
LGTM, being able to test LK + CUDA with CI will be great! Thanks @maliasadi

**Pull request merged**